src/hivpy/demographics.py

```diff
@@ -1,19 +1,149 @@
+                                0.047, 0.036, 0.027, 0.021, 0.016, 0.012])
+    stepwise_model2 = np.array([0.15, 0.13, 0.12, 0.11, 0.1, 0.09, 0.08, 0.065,
+                                0.048, 0.04, 0.03, 0.021, 0.016])
+    stepwise_model3 = np.array([0.128, 0.119, 0.113, 0.104, 0.097, 0.09, 0.081,
+                                0.074, 0.06, 0.05, 0.038, 0.026, 0.02])
+    stepwise_boundaries = np.array([-68, -55, -45, -35, -25, -15, -5, 5, 15,
+                                    25, 35, 45, 55, 65])
+
+    def __init__(self, ages, probabilities):
+        self.probability_list = probabilities
+        self.age_boundaries = ages
+
+    @classmethod
+    def selectModel(cls, inc_cat):
+        if(inc_cat == 1):
+            return cls(cls.stepwise_boundaries, cls.stepwise_model1)
+        elif(inc_cat == 2):
+            return cls(cls.stepwise_boundaries, cls.stepwise_model1)
+        else:
+            return cls(cls.stepwise_boundaries, cls.stepwise_model1)
+
+    # Generate cumulative probabilites from stepwise distribution
+    def _genCumulativeProb(self):
+        N = self.probability_list.size
+        print("IS THIS EVEN WORKING")
+        if(not np.isclose(sum(self.probability_list), 1.0, atol=1e-10, equal_nan=False)):
+            warnings.warn("Age probability distribution does not sum to one.")
+        CP = np.array([0.0]*N)
+        CP[0] = self.probability_list[0]
+        for i in range(1, N):
+            CP[i] = CP[i-1] + self.probability_list[i]
+            if(CP[i] > 1.0):
+                warnings.warn("Cumulative probability grown too large; capping at 1.0.")
+                CP[i] = 1.0
+        # make sure CP is exactly 1.0 in case of numerical error
+        CP[N-1] = 1.0
+        return CP
+
+    # Generate Age With Stepwise Distribution
+    # the size of ages should be one larger than the cumulative probabilty
+    # so that each bucket has an upper and lower bound.
+    def genAges(self, N):
+        cpd = self._genCumulativeProb()
+        p0 = 0.0
+        rands = np.random.rand(N)
+        ages_out = np.array([0.0]*N)
+        for i in range(0, cpd.size):
+            # in order to vectorise this method we create a mask of values
+            # that need to change for each age group
+            prob_mask = (p0 < rands) & (rands <= cpd[i])
+            ages_out += ((self.age_boundaries[i]
+                          + (rands - p0)/(cpd[i]-p0)
+                          * (self.age_boundaries[i+1]-self.age_boundaries[i]))
+                         * prob_mask)
+            p0 = cpd[i]
+
+        return ages_out
+
+
+class ContinuousAgeDistribution:
+    """
+    Class to handle age distributions using a continuous probability density.
+    Linear-Exponential function is used for the example distribution.
+
+    y = (mx + c)*exp(A(x-B))
+
+    This fits all three example cases quite neatly.
+    """
+
+    def _integratedLinexp(self, x, m, c, A, B):
+        return np.exp(A*(x-B))*(m*x+c - m/A)/A
```

Comment:
The docstring describes the probability density function, because that is the most intuitive to work with and select a model for. But we don't actually need the probability density to do the calculation, we need the cumulative distribution i.e. the integral of the probability density. This function is the integral (hence integratedLinexp) rather than the probability density itself. The probability density function doesn't actually appear in this class explicitly because we never call it.

Reply:
Right, that makes sense, I'd missed that.